sound/soc/sof/ipc4-pcm.c

```diff
@@ -940,20 +959,33 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 	 * Firmware calculates correct stream_start_offset for all cases including above two.
 	 * Driver subtracts stream_start_offset from dai dma position to get accurate one
 	 */
-	tmp_ptr -= time_info->stream_start_offset;
 
-	/* Calculate the delay taking into account that both pointer can wrap */
-	div64_u64_rem(tmp_ptr, substream->runtime->boundary, &tmp_ptr);
+	/*
+	 * If the dai position has passed the start offset it means that the
+	 * audio left the DSP and we need to normalize the dai position
+	 * Otherwise the first audio frames is still within the DSP, we should
+	 * adjust the host side position to accurately report the initial delay
+	 */
+	if (tmp_ptr >= time_info->stream_start_offset) {
+		tmp_ptr -= time_info->stream_start_offset;
+	} else {
+		host_ptr += time_info->stream_start_offset - tmp_ptr;
+		tmp_ptr = 0;
+	}
+
+	/* Wrap the dai position at the boundary where the host position wraps */
+	div64_u64_rem(tmp_ptr, time_info->boundary, &tmp_ptr);
+
 	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
-		head_ptr = substream->runtime->status->hw_ptr;
+		head_ptr = host_ptr;
 		tail_ptr = tmp_ptr;
 	} else {
 		head_ptr = tmp_ptr;
-		tail_ptr = substream->runtime->status->hw_ptr;
+		tail_ptr = host_ptr;
 	}
 
 	if (head_ptr < tail_ptr)
-		return substream->runtime->boundary - tail_ptr + head_ptr;
+		return time_info->boundary - tail_ptr + head_ptr;
```

Comment:
does this work for both capture and playback?

Reply:
we select the head and tail previously based on direction and the the delay is the difference, if needs to be wrapped, it is the same thing for the two direction.